Solution to Assignment 2.1 updated

divyanshjr opened the pull request:
I've tried to implement Client-Server interaction using Java. Please review and suggest changes.

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Client.java:
```diff
@@ -0,0 +1,64 @@
+import java.net.*;
+
+public class Client {
+  int self_computed_result; // To verify received result
```
Based on the rules about naming in section 5.2 of the [style guide](https://google.github.io/styleguide/javaguide.html), you be be using lowerCamelCase, not underscores.

divyanshjr replied:
Will change..

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Client.java:
```diff
@@ -0,0 +1,64 @@
+public class Client {
+  int self_computed_result; // To verify received result
+  String ip_address = "localhost"; // IP of server
+  int port = 9999;  // Port on which server is hosted
```
Extra space before `//`

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Client.java:
```diff
@@ -0,0 +1,64 @@
+  // Constructor
+  public Client() {
+    try {
+      socket = new Socket(ip_address,port);
```
Space after `,` (section 4.6.2)

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Client.java:
```diff
@@ -0,0 +1,64 @@
+    try {
+      socket = new Socket(ip_address,port);
+    }
+    catch(Exception e) {
```
Put the catch on the previous line?

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Client.java:
```diff
@@ -0,0 +1,64 @@
+
+public class Client {
+  int self_computed_result; // To verify received result
+  String ip_address = "localhost"; // IP of server
```
This isn't an IP address.

divyanshjr replied:
I'll change it to 127.0.0.1

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Client.java:
```diff
@@ -0,0 +1,64 @@
+  }
+  // Main method
+  public static void main(String args[]) throws Exception {
+    Client client = new Client();
```
Isn't it better to specify the hostname / port as constructor arguments? Make things generic?

divyanshjr replied:
Sure.

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Client.java:
```diff
@@ -0,0 +1,64 @@
+      socket = new Socket(ip_address,port);
+    }
+    catch(Exception e) {
+      System.out.println("Connection refused by server or no server found");
```
Prefer to throw Exceptions till the main function, and let it handle graceful termination.

divyanshjr replied:
Yes, that would be better. I'll do it.

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Client.java:
```diff
@@ -0,0 +1,64 @@
+    Client client = new Client();
+    client.computeResult(args);
+    client.sendData(args);
+    client.getData();
```
This method is not properly named. `getData` suggests that I'm going to receive some data as input.
However, the function returns void, and also does validation inside it.

Instead of mixing framework + application logic like this, you want to keep things separate. That way, your Client class can be used for more than exactly 1 thing.

In this case, how about if the details about the addition are isolated to your main function, and nothing else. Alternatively, don't have `main` functions in either of your classes. Create a 3rd class that uses these 2, which invokes both and ensures that communication works.

divyanshjr replied:
Okay, I'll make sure that application logic and framework are dealt separately. 

Do you mean to say that I should have a common main method for both Client and Server classses?

kaustubh-karkare replied:
Yeah, so when I coded this to get a sense of how much time it would take, I had the server and client running on separate threads, communicating via a random socket (instead of a fixed one like 9999).

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Server.java:
```diff
@@ -0,0 +1,64 @@
+import java.net.*;
+
+public class Server {
+  static ServerSocket serversocket;
```
Use camelCase to make it easier to notice word boundaries. `serverSocket`

divyanshjr replied:
Okay

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Server.java:
```diff
@@ -0,0 +1,64 @@
+    try {
+      serversocket = new ServerSocket(9999); // Trying to setup server on port 9999
+    }
+    catch(Exception e) {
```
Same as before: let `main` handle exceptions, instead of your individual methods doing it.

divyanshjr replied:
Alright

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Server.java:
```diff
@@ -0,0 +1,64 @@
+  }
+  public void start() throws Exception {
+    System.out.println("Server started");
+    while(true) {
```
Space after while.

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Client.java:
```diff
@@ -0,0 +1,64 @@
+      System.out.println("Message from server side: " + input_result);
+    }
+    else {
+        System.out.println("Result received from server: " + input_result);
```
Instead of printing directly to stdout, you should use a Logger class that can be configured.

divyanshjr replied:
I didn't know about Logger class earlier. I'll implement it.

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Server.java:
```diff
@@ -0,0 +1,64 @@
+      System.out.println("Client connected");
+      Thread.sleep(2000); // Wait for 2 seconds
+      getInput();
+      sendOutput();
```
Both of these functions need better names. `getInput` sounds like it would return something. `sendOutput` sounds like it should take an argument. Try and make names and behavior intuitive.

divyanshjr replied:
Yes, I'll change their names.

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Server.java:
```diff
@@ -0,0 +1,64 @@
+    OutputStreamWriter osw = new OutputStreamWriter(socket.getOutputStream());
+    PrintWriter out = new PrintWriter(osw);
+    // In case less or more arguments are received than required
+    if(input.length != 2) {
```
Space after `if`.

kaustubh-karkare commented on solutions/divyansh-jamuaar/Assignment-2.1/Server.java:
```diff
@@ -0,0 +1,64 @@
+      try {
+        int number1 = Integer.parseInt(input[0]);
+        int number2 = Integer.parseInt(input[1]);
+        int ans = number1 + number2;
```
This is application logic leaking into framework code again.
A method called `getOutput` shouldn't have anything to do with addition like this.

divyanshjr replied:
Alright

divyanshjr commented:
I've updated the solution to Assignment 2.1. Brief working of my code - 
-> The program now supports communication between a single server and multiple clients which run as 
    separate threads.
-> The Server has an additional Worker thread (ClientServiceThread) which processes the requests of the 
    clients. The Server accepts requests from the clients and passes them to the worker thread which 
    processes them and respnds back to the respective client.
-> All the messages generated during the communication are stored in a Log File. This feature is 
     implemented with the help of Logger Class in Java.

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+// Class containing the main method
+class Assignment2_1 {
+  private final static Logger logger = Logger.getLogger(Assignment2_1.class.getName());
+  private static FileHandler fh = null;
```
No abbreviations please.

divyanshjr replied:
oops.. need to change my habit.. :P

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+ *Multithreading approach is used where Server and Clients run as separate threads
+ *Server has an additional ClientServiceThread which processes the requests of the clients
+ *The Server accepts requests from Clients and passes the requests to ClientServiceThread
+ *All messages are recorded in a LOG File
```
According to the style guide, you should have spaces after `*`.

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+  public Server(int port) throws Exception {
+      Server.port = port;
+      serverSocket = new ServerSocket(port);
+    }
```
The indenting for the last 3 lines is off.

divyanshjr replied:
I'll correct that.

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+    }
+  @Override
+  public void run() {
+    logger.log(Level.INFO, "Server Started");
```
I dont know the exact API, but based on how other languages do it, I would expect that there will be methods like `logger.info` here. Could you confirm?

divyanshjr replied:
logger.info() takes just one parameter i.e. the message and sets it to the level INFO. logger.log() is a generalised method which takes 2 arguments (Level , Message). I could even use methods like logger.info(), logger.warning() etc if you wish.

kaustubh-karkare replied:
Prefer `logger.info`. This looks unnecessarily verbose.

divyanshjr replied:
Alright.. :)

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+      try {
+        socket = serverSocket.accept();
+      } catch(Exception e) {}
+      logger.log(Level.INFO, "Client connected");
```
Specify information about the client?

divyanshjr replied:
Alright.

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+// Class which is responsible for processing Client request
+class ClientServiceThread extends Thread {
+  Socket socket;
+  //Constructor
```
nit: Space after `//`

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+  public void run() {
+    try {
+      Thread.sleep(2000); // Wait for 2 seconds
+    } catch(Exception e) {}
```
In general, catch-all exceptions are a bad idea, since they hide information that could otherwise be useful.
In fact, I'd rather not have exceptions caught at all, unless it is information that you actively use.
What's a valid usecase of catching? Database connection failure, since that is something you can retry.
In most cases, just let it bubble up to `main`, which can then figure out what to do.

divyanshjr replied:
Agree, I'll try to specify the Exception type in the catch statement. But here I won't be able to propagate my Exception to the main method since the run() method is not allowed to throw any exception.

kaustubh-karkare replied:
Hmm. It seems weird to have 2 consecutive statements like this. I get why you're doing it, but still, looks weird :P

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+    try {
+      String[] input; // Input received from Client
+      String reply; // Reply to be sent to Client
+      input = fetchClientRequest();
```
nit: Use `get` instead of `fetch`? Fetch usually suggests that you have to make a network call to obtain the data, which get is for already available data. This is borderline, but the data is already delivered, so `get` makes more sense to me.

divyanshjr replied:
Sure. I'll change it.

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+      String reply; // Reply to be sent to Client
+      input = fetchClientRequest();
+      reply = processData(input);
+      replyClient(reply);
```
`replyToClient` might be better.

divyanshjr replied:
Alright.

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+      String[] input; // Input received from Client
+      String reply; // Reply to be sent to Client
+      input = fetchClientRequest();
+      reply = processData(input);
```
It's cool that you have isolated the adding logic to this one method. Let's take it one step further, and take it is an input to your `Server` class.

divyanshjr replied:
I didn't get you. Could you please explain a bit more?

kaustubh-karkare replied:
Right now, the logic for addition is hardcoded in `processData`. But we want to make it even more flexible, and allow the users of this class to specify how to interact with the client. Basically, make it work beyond this one exact example (you'll hear be say this a lot).

divyanshjr replied:
yeah.. it is.. but I'll try..

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+    } catch(Exception e) {}
+  }
+  // Method to fetch data from client
+  String[] fetchClientRequest() throws Exception {
```
I'm not sure it makes sense to even have this method (and replyClient).
This design works when the client makes one request, and gets one response, but in order to make things truly generic, you cant make that assumption in the general case.
I'd think you could store your ObjectInputStream and OutputStreamWriter as class properties, and use them directly from `run`.

divyanshjr replied:
Okay, so that means I should only have a processData() method other than run() and rest of the job should be done inside the run() method.

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+        reply = Integer.toString(ans);
+      }
+      // In case input arguments are not of integer type
+      catch(Exception e) {
```
Your comment helps, but you should use the proper exception corresponding to that case.

divyanshjr replied:
Yes, I'll catch the specific Exception only.

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+  // Constructor
+  public Client(String IP, int port, String[] input, int clientID) throws Exception {
+    socket = new Socket(IP, port);
+    arguments = input;
```
This is confusing. Would prefer to have a single name for a given piece of data consistently.

divyanshjr replied:
Sure.

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+    validate(response);
+  }
+  // Method to send client data to server
+  public void queryServer() throws Exception {
```
My comments about `fetchClientRequest` and `replyClient` apply here too.

divyanshjr replied:
Agree, So only one method should be there to validate the response. Rest of the job is to be done inside run().

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+    return response;
+  }
+  // Method to validate the answer recieved from server
+  public void validate(String response) {
```
This method is a bit hard to read. Can you simplify?

divyanshjr replied:
I'll try to simplify.

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+        logger.log(Level.FINE, "Client "+clientID+": Result verified");
+      }
+      else {
+        logger.log(Level.WARNING, "Client "+clientID+": Incorrect result obtained");
```
According to style guide, spaces around operators.

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+    catch (SecurityException | IOException e) {
+      e.printStackTrace();
+    }
+    Logger l = Logger.getLogger("");
```
Same as above.

kaustubh-karkare commented on solutions/Assignment2_1.java:
```diff
@@ -0,0 +1,206 @@
+    clientsCount = scanner.nextInt();
+    client = new Client[clientsCount];
+    // Trying to set up the server
+    System.out.println("Enter port on which server has to be hosted");
```
Use command line arguments for this. It is okay to use an existing options parser library this time :)

divyanshjr replied:
Alright. I'll do it.
